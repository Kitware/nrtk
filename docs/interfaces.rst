##########
Interfaces
##########

The NRTK API consists of a number of object-oriented functor interfaces for item-response curve
(IRC) generation,
namely for assessing model response to perturbations on given input data. These interfaces focus on black-box IRC
generation. In addition to the driver, or *generator*, of this task, there are two other main components: reference
image perturbation in preparation for black-box testing and black-box scoring for model outputs. The
``PeturbImageFactory`` interface provides the utility to easily vary specified parameters of a particular perturber.
The generator will execute upon a given perturber factory as well as a given model and scorer to generate the IRC for
given input data. We define a few similar interfaces for performing the IRC generation, separated by the intermediate
algorithmic use cases, one for object detection and one for image classification.


We explicitly do not require an abstraction for the black-box operations to fit inside. This is intended to allow for
applications using these interfaces while leveraging existing functionality, which only need to perform data formatting
to fit the input defined here. Note, however, some interfaces are defined for certain black-box concepts as part of the
SMQTK ecosystem (e.g. in `SMQTK-Classifier <https://github.com/Kitware/SMQTK-Classifier>`_,
`SMQTK-Detection
<https://github.com/Kitware/SMQTK-Detection>`_, and other SMQTK-* modules).


These interfaces are based on the plugin and configuration features provided by
`SMQTK-Core <https://github.com/Kitware/SMQTK-Core>`_, to allow convenient hooks into implementation, discoverability,
and factory generation from runtime configuration. This allows for both opaque discovery of interface implementations
from a class-method on the interface class object, as well as instantiation of a concrete instance via a JSON-like
configuration fed in from an outside resource.

.. figure:: figures/api-docs-fig-01.svg

   Figure 1: Abstract Interface Inheritance.

.. When adding new classes within interfaces, sort them alphabetically.

------------------
Image Perturbation
------------------

Interface: PerturbImage
-----------------------
.. autoclass:: nrtk.interfaces.perturb_image.PerturbImage
   :members:
   :special-members:

--------------------
Perturbation Factory
--------------------

Interface: PerturbImageFactory
------------------------------
.. autoclass:: nrtk.interfaces.perturb_image_factory.PerturbImageFactory
   :members:
   :special-members:

-------------
Image Metrics
-------------

Interface: ImageMetric
----------------------
.. autoclass:: nrtk.interfaces.image_metric.ImageMetric
   :members:
   :special-members:

-------
Scoring
-------

Interface: ScoreDetections
--------------------------
.. autoclass:: nrtk.interfaces.score_detections.ScoreDetections
   :members:
   :special-members:

---------------------------------
End-to-End Generation and Scoring
---------------------------------

Interface: GenerateObjectDetectorBlackboxResponse
-------------------------------------------------
.. autoclass:: nrtk.interfaces.gen_object_detector_blackbox_response.GenerateObjectDetectorBlackboxResponse
   :members:
   :special-members:
