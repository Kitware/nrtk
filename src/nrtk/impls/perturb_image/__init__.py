"""Module for all PerturbImage implementations"""
