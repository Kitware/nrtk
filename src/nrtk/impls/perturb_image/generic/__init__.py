"""Module for all generic implementations of PerturbImage"""
