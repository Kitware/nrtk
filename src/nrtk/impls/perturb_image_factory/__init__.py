"""Module for all implementations of PerturbImageFactory"""
