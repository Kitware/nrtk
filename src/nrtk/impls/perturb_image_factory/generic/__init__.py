"""Module for generic implementations of PerturbImageFactory"""
