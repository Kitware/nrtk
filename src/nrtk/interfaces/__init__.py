"""package housing the interfaces of nrtk"""
