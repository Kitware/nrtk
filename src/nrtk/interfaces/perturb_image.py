import abc
from typing import Any, Dict, Optional

import numpy as np
from smqtk_core import Plugfigurable


class PerturbImage(Plugfigurable):
    """Algorithm that generates a perturbed image for given input image stimulus as a ``numpy.ndarray`` type array."""

    @abc.abstractmethod
    def perturb(self, image: np.ndarray, additional_params: Optional[Dict[str, Any]] = None) -> np.ndarray:
        """Generate a perturbed image for the given image stimulus.

        Note perturbers that resize, rotate, or similarly affect the dimensions of an image may impact
        scoring if bounding boxes are not similarly transformed.

        :param image: Input image as a numpy array.
        :param additional_params: A dictionary containing perturber implementation-specific input param-values pairs.

        :return: Perturbed image as numpy array, including matching dtype. Implementations should impart no side
            effects upon the input image.
        """
        if additional_params is None:
            additional_params = dict()
        return image

    def __call__(self, image: np.ndarray, additional_params: Optional[Dict[str, Any]] = None) -> np.ndarray:
        """Calls ``perturb()`` with the given input image."""
        if additional_params is None:
            additional_params = dict()
        return self.perturb(image, additional_params)

    @classmethod
    def get_type_string(cls) -> str:
        return f"{cls.__module__}.{cls.__name__}"
