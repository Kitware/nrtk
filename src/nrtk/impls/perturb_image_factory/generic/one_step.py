from typing import Any, Dict, Type

from nrtk.impls.perturb_image_factory.generic.step import StepPerturbImageFactory
from nrtk.interfaces.perturb_image import PerturbImage


class OneStepPerturbImageFactory(StepPerturbImageFactory):
    """Simple PerturbImageFactory implementation to return a factory with one perturber."""

    def __init__(
        self,
        perturber: Type[PerturbImage],
        theta_key: str,
        theta_value: float,
    ):
        """Initialize the factory to produce an instance of PerturbImage for the given type.

        Initialize the factory to produce an instance of PerturbImage for the given type,
        given the ``theta_key`` and the ``theta_value`` parameters.

        :param perturber: Python implementation type of the PerturbImage interface
            to produce.

        :param theta_key: Perturber parameter to set for the instance.

        :param theta_value: Initial and only value of ``theta_key``.

        :raises TypeError: Given a perturber instance instead of type.
        """
        super().__init__(perturber=perturber, theta_key=theta_key, start=theta_value, stop=theta_value + 0.1, step=1.0)

        self.theta_value = theta_value

    def get_config(self) -> Dict[str, Any]:
        cfg = super().get_config()
        cfg["theta_value"] = self.theta_value
        return {k: cfg[k] for k in ("perturber", "theta_key", "theta_value")}
