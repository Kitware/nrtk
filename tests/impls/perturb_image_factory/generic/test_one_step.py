import json
from collections.abc import Hashable, Iterable
from pathlib import Path
from typing import Any, Optional

import numpy as np
import pytest
from smqtk_core.configuration import (
    configuration_test_helper,
    from_config_dict,
    to_config_dict,
)
from smqtk_image_io.bbox import AxisAlignedBoundingBox
from syrupy.assertion import SnapshotAssertion

from nrtk.impls.perturb_image_factory.generic.one_step import OneStepPerturbImageFactory
from nrtk.interfaces.perturb_image import PerturbImage
from nrtk.interfaces.perturb_image_factory import PerturbImageFactory

DATA_DIR = Path(__file__).parents[3] / "data"
INPUT_IMG_FILE_PATH = "./examples/pybsm/data/M-41 Walker Bulldog (USA) width 319cm height 272cm.tiff"


class DummyPerturber(PerturbImage):
    def __init__(self, param_1: int = 1, param_2: int = 2) -> None:
        self.param_1 = param_1
        self.param_2 = param_2

    def perturb(
        self,
        image: np.ndarray,
        boxes: Optional[Iterable[tuple[AxisAlignedBoundingBox, dict[Hashable, float]]]] = None,
        additional_params: Optional[dict[str, Any]] = None,  # noqa: ARG002
    ) -> tuple[np.ndarray, Optional[Iterable[tuple[AxisAlignedBoundingBox, dict[Hashable, float]]]]]:
        return np.copy(image), boxes

    def get_config(self) -> dict[str, Any]:
        return {"param_1": self.param_1, "param_2": self.param_2}


class TestStepPerturbImageFactory:
    @pytest.mark.parametrize(
        ("perturber", "theta_key", "theta_value"),
        [
            (DummyPerturber, "param_1", 1.0),
            (DummyPerturber, "param_2", 3.0),
        ],
    )
    def test_iteration(
        self,
        snapshot: SnapshotAssertion,
        perturber: type[PerturbImage],
        theta_key: str,
        theta_value: float,
    ) -> None:
        """Ensure factory can be iterated upon and the varied parameter matches expectations."""
        factory = OneStepPerturbImageFactory(perturber=perturber, theta_key=theta_key, theta_value=theta_value)
        assert len(factory) == 1
        assert factory[0].get_config() == snapshot

    @pytest.mark.parametrize(
        ("perturber", "theta_key", "theta_value"),
        [(DummyPerturber, "param_1", 1.0), (DummyPerturber, "param_2", 3.0)],
    )
    def test_configuration(self, perturber: type[PerturbImage], theta_key: str, theta_value: float) -> None:
        """Test configuration stability."""
        inst = OneStepPerturbImageFactory(perturber=perturber, theta_key=theta_key, theta_value=theta_value)
        for i in configuration_test_helper(inst):
            assert i.perturber == perturber
            assert i.theta_key == theta_key
            assert i.theta_value == theta_value
            assert i.stop == theta_value + 0.1
            assert i.step == 1.0
            assert i.to_int
            assert i.theta_value in i.thetas
            assert i.stop not in i.thetas

    @pytest.mark.parametrize(
        ("perturber", "theta_key", "theta_value"),
        [(DummyPerturber, "param_1", 1.0), (DummyPerturber, "param_2", 3.0)],
    )
    def test_hydration(
        self,
        tmp_path: Path,
        perturber: type[PerturbImage],
        theta_key: str,
        theta_value: float,
    ) -> None:
        """Test configuration hydration using from_config_dict."""
        original_factory = OneStepPerturbImageFactory(perturber=perturber, theta_key=theta_key, theta_value=theta_value)

        original_factory_config = original_factory.get_config()

        config_file_path = tmp_path / "config.json"
        with open(str(config_file_path), "w") as f:
            json.dump(to_config_dict(original_factory), f)

        with open(str(config_file_path)) as config_file:
            config = json.load(config_file)
            hydrated_factory = from_config_dict(config, PerturbImageFactory.get_impls())
            hydrated_factory_config = hydrated_factory.get_config()

            assert original_factory_config == hydrated_factory_config
