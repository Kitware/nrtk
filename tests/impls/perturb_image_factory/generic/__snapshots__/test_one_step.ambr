# serializer version: 1
# name: TestStepPerturbImageFactory.test_iteration[DummyPerturber-param1-1.0]
  dict({
    'param1': 1,
    'param2': 2,
  })
# ---
# name: TestStepPerturbImageFactory.test_iteration[DummyPerturber-param2-3.0]
  dict({
    'param1': 1,
    'param2': 3,
  })
# ---
